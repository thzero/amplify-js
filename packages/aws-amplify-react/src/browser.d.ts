declare interface Document {
	exitFullscreen: () => void;
	mozCancelFullScreen: () => void;
	webkitExitFullscreen: () => void;
	fullscreenElement: () => void;
	mozFullScreenElement: () => void;
	webkitFullscreenElement: () => void;
}

declare interface Window {
	amazon: any;
	auth0: any;
	auth0_client: any;
	FB: any;
	fbAsyncInit: any;
	gapi: any;
}
