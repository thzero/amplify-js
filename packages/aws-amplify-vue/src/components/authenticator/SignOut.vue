/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

<template>
  <div v-bind:data-test="auth.signOut.section">
   <div v-bind:class="amplifyUI.formField">
    <!--
      <div v-bind:class="amplifyUI.inputLabel">{{options.msg}}</div>
      <button v-bind:class="amplifyUI.button" v-on:click="signOut" v-bind:data-test="auth.signOut.button">{{options.signOutButton}}</button>
      -->
      <div v-bind:class="amplifyUI.inputLabel">{{options.msg}}</div>
      <button v-bind:class="amplifyUI.button" v-on:click="signOut" v-bind:data-test="auth.signOut.button">{{options.signOutButton}}</button>
    </div>
    <div class="error" v-if="error">
      {{ error }}
    </div>
  </div>
</template>

<script>
import AmplifyEventBus from '../../events/AmplifyEventBus';
import * as AmplifyUI from '@aws-amplify/ui';
import { existsSync } from 'fs';

import constants from './common/constants.js';
import { auth } from '../../assets/data-test-attributes';

export default {
  name: 'SignOut',
  props: ['signOutConfig'],
  data () {
    return {
      error: '',
      show: false,
      amplifyUI: AmplifyUI,
      auth,
      logger: {},
    }
  },
  computed: {
    options() {
      const defaults = {
        msg: null,
        signOutButton: this.$Amplify.I18n.get('Sign Out')
      }
      return Object.assign(defaults, this.signOutConfig || {})
    }
  },
  async mounted() {
    this.logger = new this.$Amplify.Logger(this.$options.name);
  },
  methods: {
    signOut: function(event) {
      let payload = {};
      try {
        payload = JSON.parse(localStorage.getItem(constants.AUTH_SOURCE_KEY)) || {};
        localStorage.removeItem(constants.AUTH_SOURCE_KEY);
      } catch (e) {
        this.logger.debug(`Failed to parse the info from ${constants.AUTH_SOURCE_KEY} from localStorage with ${e}`);
      }

      this.logger.debug('sign out from the source', payload);
      switch (payload.provider) {
        case constants.GOOGLE:
          this.googleSignOut();
          break;
        case constants.FACEBOOK:
          this.facebookSignOut();
          break;
        case constants.AMAZON:
          this.amazonSignOut();
          break;
        default:
          break;
      }

      if (!this.$Amplify.Auth || typeof this.$Amplify.Auth.signOut !== 'function') {
          throw new Error('No Auth module found, please ensure @aws-amplify/auth is imported');
      }

      this.$Amplify.Auth.signOut()
        .then(() => {
          this.logger.info('signout success')
          return AmplifyEventBus.$emit('authState', 'signedOut')
        })
        .catch(e => this.setError(e));
    },
    amazonSignOut: function() {
      const amz = window.amazon;
      if (!amz) {
        this.logger.debug('Amazon Login sdk undefined');
        return Promise.resolve();
      }

      this.logger.debug('Amazon signing out');
      amz.Login.logout();
    },
    facebookSignOut: function() {
      const fb = window.FB;
      if (!fb) {
        this.logger.debug('FB sdk undefined');
        return Promise.resolve();
      }

      fb.getLoginStatus(response => {
        if (response.status === 'connected') {
          return new Promise((res, rej) => {
            this.logger.debug('facebook signing out');
            fb.logout(response => {
              res(response);
            });
          });
        } else {
          return Promise.resolve();
        }
      });
    },
    googleSignOut: function() {
      const ga = window.gapi && window.gapi.auth2
        ? window.gapi.auth2.getAuthInstance()
        : null;
      if (!ga) {
        return Promise.resolve();
      }

      ga.then((googleAuth) => {
        if (!googleAuth) {
          this.logger.debug('google Auth undefined');
          return Promise.resolve();
        }

        this.logger.debug('google signing out');
        return googleAuth.signOut();
      });
    },
    setError: function(e) {
      this.error = this.$Amplify.I18n.get(e.message || e);
      this.logger.error(this.error)
    }
  }
}
</script>
