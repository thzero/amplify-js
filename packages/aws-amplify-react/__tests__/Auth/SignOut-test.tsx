import Auth from '@aws-amplify/auth';
import SignOut from '../../src/Auth/SignOut';
import * as React from 'react';
import AmplifyTheme from '../../src/AmplifyTheme';
import AuthPiece from '../../src/Auth/AuthPiece';
import { Header, Footer, InputRow, ButtonRow } from '../../src/AmplifyUI';
import { Hub } from '@aws-amplify/core';
import * as AmplifyMocks from '../../__mocks__/amplifyMock';

const acceptedStates = ['signedIn'];

const deniedStates = [
	'signIn',
	'signedUp',
	'signedOut',
	'forgotPassword',
	'signUp',
	'confirmSignIn',
	'confirmSignUp',
	'verifyContact',
];

describe('SignOut', () => {
	describe('normal case', () => {
		test('render correctly with authState signedIn', () => {
			const wrapper = shallow(<SignOut />);
			for (let i = 0; i < acceptedStates.length; i += 1) {
				wrapper.setProps({
					authState: acceptedStates[i],
					theme: 'theme',
				});
				expect(wrapper).toMatchSnapshot();
			}
		});

		test('render correctly with hide', () => {
			const wrapper = shallow(<SignOut />);
			for (let i = 0; i < acceptedStates.length; i += 1) {
				wrapper.setProps({
					authState: acceptedStates[i],
					theme: 'theme',
					hide: [SignOut],
				});
				expect(wrapper).toMatchSnapshot();
			}
		});

		test('render correctly with empty hide', () => {
			const wrapper = shallow(<SignOut />);
			for (let i = 0; i < acceptedStates.length; i += 1) {
				wrapper.setProps({
					authState: acceptedStates[i],
					theme: 'theme',
					hide: [],
				});
				expect(wrapper).toMatchSnapshot();
			}
		});
	});

	test('render correctly with other authStates', () => {
		const wrapper = shallow(<SignOut />);

		for (let i = 0; i < deniedStates.length; i += 1) {
			wrapper.setProps({
				authState: deniedStates[i],
				theme: 'theme',
			});

			expect(wrapper).toMatchSnapshot();
		}
	});

	describe('signOut test', () => {
		test('happy case (without stateFromStorage)', async () => {
			const wrapper = shallow(<SignOut />);
			const signOut = wrapper.instance();

			const spyon = jest.spyOn(Auth, 'signOut').mockImplementationOnce(() => {
				return Promise.resolve();
			});

			await signOut.signOut();

			expect(spyon).toBeCalled();
			spyon.mockClear();
		});

		test('happy case (with stateFromStorage)', async () => {
			const wrapper = shallow(<SignOut />);
			const signOut = wrapper.instance();

			signOut.state.stateFromStorage = true;

			const spyon = jest.spyOn(Auth, 'signOut').mockImplementationOnce(() => {
				return Promise.resolve();
			});

			await signOut.signOut();

			expect(spyon).toBeCalled();
			spyon.mockClear();
		});

		test('error case', async () => {
			const wrapper = shallow(<SignOut />);
			const signOut = wrapper.instance();

			const spyon = jest.spyOn(Auth, 'signOut').mockImplementationOnce(() => {
				return Promise.reject('error');
			});

			await signOut.signOut();

			expect(spyon).toBeCalled();
			spyon.mockClear();
		});
	});

	describe('onHubCapsule tests', () => {
		test('onHubCapsule is present', () => {
			const signOut = mount(<SignOut />).instance();
			expect(signOut.onHubCapsule).toBeTruthy();
		});
		test('onHubCapsule is called on a Hub event', () => {
			const spy = jest.spyOn(SignOut.prototype, 'onHubCapsule');
			mount(<SignOut />).instance();
			Hub.dispatch('auth', { event: 'test' });
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
		test('onHubCapsule should setState with authState = "signedIn" when "signIn" auth event fires', () => {
			const spy = jest.spyOn(SignOut.prototype, 'setState');
			mount(<SignOut />).instance();
			Hub.dispatch('auth', { event: 'signIn', data: { foo: 'bar' } });
			expect(spy).toHaveBeenCalledWith({
				authState: 'signedIn',
				authData: { foo: 'bar' },
			});
			spy.mockClear();
		});
		test('onHubCapsule should setState with authState = "signIn" when "customSignOut" auth event fires', () => {
			const spy = jest.spyOn(SignOut.prototype, 'setState');
			mount(<SignOut />).instance();
			Hub.dispatch('auth', { event: 'signOut' });
			expect(spy).toHaveBeenCalledWith({
				authState: 'signIn',
			});
			spy.mockClear();
		});
	});

	describe('findState tests', () => {
		test('findState is called', () => {
			const spy = jest.spyOn(SignOut.prototype, 'findState');
			mount(<SignOut />).instance();
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
		test('Auth.currentAuthenticatedUser is not called if auth props are present', () => {
			const spy = jest
				.spyOn(Auth, 'currentAuthenticatedUser')
				.mockImplementationOnce(() => {
					return new Promise((res, rej) => {
						res({
							user: {},
						});
					});
				});
			const props = {
				authState: 'signedIn',
				authData: {},
			};
			const wrapper = shallow(<SignOut {...props} />);
			wrapper.instance();
			expect(spy).not.toHaveBeenCalled();
			spy.mockClear();
		});
		test('Auth.currentAuthenticatedUser is called if auth props are not present', () => {
			const spy = jest
				.spyOn(Auth, 'currentAuthenticatedUser')
				.mockImplementationOnce(() => {
					return new Promise((res, rej) => {
						res({
							user: {},
						});
					});
				});
			const props = {};
			const wrapper = shallow(<SignOut {...props} />);
			wrapper.instance();
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
		test('Auth.currentAuthenticatedUser results in state being set', async () => {
			const spy = jest.spyOn(SignOut.prototype, 'setState');
			const props = {};
			const wrapper = shallow(<SignOut {...props} />);
			const signOut = wrapper.instance();
			await signOut.findState();
			expect(signOut.state.stateFromStorage).toEqual(true);
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
	});

	describe('SignOut lifecycle', () => {
		test('componentDidMount', async () => {
			const spy = jest.spyOn(SignOut.prototype, 'componentDidMount');
			const signOut = mount(<SignOut />).instance();
			signOut.componentDidMount();
			expect(signOut._isMounted).toBeTruthy();
			expect(spy).toHaveBeenCalled();
			spy.mockClear();
		});
		test('componentWillUnmount', async () => {
			const wrapper = shallow(<SignOut />);
			const signOut = wrapper.instance();
			const componentWillUnmount = jest.spyOn(signOut, 'componentWillUnmount');
			wrapper.unmount();
			expect(signOut._isMounted).toBeFalsy();
			expect(componentWillUnmount).toHaveBeenCalled();
		});
	});
});
