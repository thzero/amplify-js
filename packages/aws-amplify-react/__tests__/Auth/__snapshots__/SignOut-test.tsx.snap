// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignOut normal case render correctly with authState signedIn 1`] = `
<Component
  data-test="sign-out-button"
  onClick={[Function]}
  theme="theme"
>
  Sign Out
</Component>
`;

exports[`SignOut normal case render correctly with empty hide 1`] = `
<Component
  data-test="sign-out-button"
  onClick={[Function]}
  theme="theme"
>
  Sign Out
</Component>
`;

exports[`SignOut normal case render correctly with hide 1`] = `""`;

exports[`SignOut render correctly with other authStates 1`] = `""`;

exports[`SignOut render correctly with other authStates 2`] = `""`;

exports[`SignOut render correctly with other authStates 3`] = `""`;

exports[`SignOut render correctly with other authStates 4`] = `""`;

exports[`SignOut render correctly with other authStates 5`] = `""`;

exports[`SignOut render correctly with other authStates 6`] = `""`;

exports[`SignOut render correctly with other authStates 7`] = `""`;

exports[`SignOut render correctly with other authStates 8`] = `""`;
